### PR TITLE
Spanner Samples

### PR DESCRIPTION
**Pull request by remi**

### CODE REVIEW

File: spanner/spanner_samples.rb [hunk 1]
```diff
@@ -0,0 +1,391 @@
...
+  # project_id  = "Your Google Cloud project ID"
+  # instance_id = "Your Spanner instance ID"
+  # database_id = "Your Spanner database ID"
+
+  require "google/cloud/spanner"
+
+  spanner = Google::Cloud::Spanner.new project: project_id
+  client  = spanner.client instance_id, database_id
+
+  client.transaction do |tx|
```
Comment (frankyn) [inline spanner/spanner_samples.rb]: How about tx to transaction?

File: spanner/spanner_samples.rb [hunk 2]
```diff
@@ -0,0 +1,391 @@
...
+  read_only_transaction        <instance_id> <database_id> Read-Only Transaction
+
+Environment variables:
+  GOOGLE_CLOUD_PROJECT must be set to your Google Cloud project ID
+    usage
+end
+
+def run_sample arguments
+  command     = arguments.shift
+  project_id  = ENV["GOOGLE_CLOUD_PROJECT"]
```
Comment (frankyn) [inline spanner/spanner_samples.rb]: Nit* (you can ignore): I'd like to have ENV before or after arguments.shift for grouping.

File: spanner/spec/spanner_samples_spec.rb [hunk 3]
```diff
@@ -0,0 +1,452 @@
...
+    )
+    expect(captured_output).to include(
+      "Created database #{@database_id} on instance #{@instance.instance_id}"
+    )
+
+    database = @instance.database @database_id
+    expect(database).not_to be nil
+
+    data_definition_statements = database.ddl
+    expect(data_definition_statements.size).to eq 2
```
Comment (frankyn) [inline spanner/spec/spanner_samples_spec.rb]: align eq with include on the following line.

File: spanner/spec/spanner_samples_spec.rb [hunk 4]
```diff
@@ -0,0 +1,452 @@
...
+    expect(data_definition_statements.first).to include "CREATE TABLE Singers"
+    expect(data_definition_statements.last).to  include "CREATE TABLE Albums"
+  end
+
+  example "insert data" do
+    database = create_singers_albums_database
+    client   = @spanner.client @instance.instance_id, database.database_id
+
+    expect(client.execute("SELECT * FROM Singers").rows.count).to eq 0
+    expect(client.execute("SELECT * FROM Albums").rows.count).to eq 0
```
Comment (frankyn) [inline spanner/spec/spanner_samples_spec.rb]: align eq with the previous line.

File: spanner/spec/spanner_samples_spec.rb [hunk 5]
```diff
@@ -0,0 +1,452 @@
...
+                             instance_id: @instance.instance_id,
+                             database_id: database.database_id
+    end
+
+    expect(captured_output).to include "Transaction complete"
+
+    first_album  = client.read("Albums", [:MarketingBudget], keys: [[1,1]]).rows.first
+    second_album = client.read("Albums", [:MarketingBudget], keys: [[2,2]]).rows.first
+
+    expect(first_album[:MarketingBudget]).to eq 300_000
```
Comment (frankyn) [inline spanner/spec/spanner_samples_spec.rb]: align eq with following line.

File: spanner/spanner_samples.rb [hunk 6]
```diff
@@ -0,0 +1,391 @@
...
+  # database_id = "Your Spanner database ID"
+
+  require "google/cloud/spanner"
+
+  spanner  = Google::Cloud::Spanner.new project: project_id
+  instance = spanner.instance instance_id
+  database = instance.database database_id
+
+  job = database.update statements: [
+    "CREATE INDEX AlbumsByAlbumTitle2 ON Albums(AlbumTitle) " +
```
Comment (frankyn) [inline spanner/spanner_samples.rb]: Why not remove the `" +`? Referencing line 27, it's a multi-line constant string.

File: spanner/spanner_samples.rb [hunk 7]
```diff
@@ -0,0 +1,390 @@
...
+  # instance_id = "Your Spanner instance ID"
+  # database_id = "Your Spanner database ID"
+
+  require "google/cloud/spanner"
+
+  spanner = Google::Cloud::Spanner.new project: project_id
+  client  = spanner.client instance_id, database_id
+
+  client.transaction do |tx|
+    first_album  = tx.read("Albums", [:MarketingBudget], keys: [[1,1]]).rows.first
```
Comment (frankyn) [inline spanner/spanner_samples.rb]: keys: [[1,1]] is ambiguous to me when I first read this sample.

File: spanner/spanner_samples.rb (hunk 2 above)
Comment (remi) [inline spanner/spanner_samples.rb]: This comment makes me immensely happy

Comment (remi) [inline spanner/spanner_samples.rb]: ( Also: fixed! )

File: spanner/spec/spanner_samples_spec.rb (hunk 3 above)
Comment (remi) [inline spanner/spec/spanner_samples_spec.rb]: 🎉 Prettier now! 😎

File: spanner/spanner_samples.rb [hunk 8]
```diff
@@ -0,0 +1,390 @@
...
+  # instance_id = "Your Spanner instance ID"
+  # database_id = "Your Spanner database ID"
+
+  require "google/cloud/spanner"
+
+  spanner = Google::Cloud::Spanner.new project: project_id
+  client  = spanner.client instance_id, database_id
+
+  client.execute("SELECT SingerId, AlbumId, AlbumTitle FROM Albums").rows.each do |row|
+    puts "#{row[:SingerId]} #{row[:AlbumId]} #{row[:AlbumTitle]}"
```
Comment (frankyn) [inline spanner/spanner_samples.rb]: Symbols are great! Is it assumed that a symbol exists for each column? I'd appreciate having this called out in the documentation for Ruby.

File: spanner/spanner_samples.rb [hunk 9]
```diff
@@ -0,0 +1,390 @@
...
+  # database_id = "Your Spanner database ID"
+
+  require "google/cloud/spanner"
+
+  spanner = Google::Cloud::Spanner.new project: project_id
+  client  = spanner.client instance_id, database_id
+
+  client.commit do |c|
+    c.insert "Singers", [
+      { SingerId: 1, FirstName: "Marc",     LastName: "Richards" },
```
Comment (frankyn) [inline spanner/spanner_samples.rb]: You aligned `{ }` in other locations. Should this be consistent here? Maybe not because it would leave a lot of whitespace.

File: spanner/spanner_samples.rb (hunk 6 above)
Comment (remi) [inline spanner/spanner_samples.rb]: The server didn't like the SQL query with tons of extra whitesace when I tried that - let me try again now...

Comment (remi) [inline spanner/spanner_samples.rb]: Fixed

File: spanner/spanner_samples.rb (hunk 9 above)
Comment (remi) [inline spanner/spanner_samples.rb]: Fixed

File: spanner/spanner_samples.rb (hunk 8 above)
Comment (remi) [inline spanner/spanner_samples.rb]: (discussion offline)

File: spanner/spanner_samples.rb [hunk 10]
```diff
@@ -0,0 +1,390 @@
...
+  # project_id  = "Your Google Cloud project ID"
+  # instance_id = "Your Spanner instance ID"
+  # database_id = "Your Spanner database ID"
+
+  require "google/cloud/spanner"
+
+  spanner = Google::Cloud::Spanner.new project: project_id
+  client  = spanner.client instance_id, database_id
+
+  result = client.read "Albums", [:AlbumId, :AlbumTitle, :SingerId],
```
Comment (frankyn) [inline spanner/spanner_samples.rb]: Why not use [snapshot](https://googlecloudplatform.github.io/google-cloud-ruby/#/docs/google-cloud/v0.35.0/google/cloud/spanner/client?method=snapshot-instance)? IIRC, you mentioned other languages use this higher level method for read only transactions. Might make this sample easier to read!

Comment (remi) [inline spanner/spanner_samples.rb]: Ack. I will update code and send you the docs snippet for this.

Comment (remi) [inline spanner/spanner_samples.rb]: (Thanks! When I was coding this, I somehow missed this. This is exactly what I want!)

File: spanner/spec/spanner_samples_spec.rb [hunk 11]
```diff
@@ -0,0 +1,452 @@
...
+    expect(captured_output).to include "Waiting for database update to complete"
+    expect(captured_output).to include "Added the AlbumsByAlbumTitle index"
+
+    expect(database.ddl(force: true).join).to include(
+      "CREATE INDEX AlbumsByAlbumTitle ON Albums(AlbumTitle)"
+    )
+  end
+
+  example "create storing index" do
+    database = create_singers_albums_database
```
Comment (frankyn) [inline spanner/spec/spanner_samples_spec.rb]: Database is missing the MarketingBudget column and that's why the expectation on line 190 fails. It passes for me once I added this column. 

```ruby
 add_column project_id:    @project_id,
                      instance_id:  @instance.instance_id,
                      database_id: database.database_id
```

Comment (remi) [inline spanner/spec/spanner_samples_spec.rb]: Fixed.

**Comment by frankyn** [comment]
LGTM
